avm-transpiler/src/transpile.rs

```diff
@@ -228,9 +233,121 @@ fn handle_foreign_call(
     function: &String,
     destinations: &Vec<ValueOrArray>,
     inputs: &Vec<ValueOrArray>,
+) {
+    match function.as_str() {
+        "keccak256" | "sha256" => {
+            handle_2_field_hash_instruction(avm_instrs, function, destinations, inputs)
+        }
+        "poseidon" => handle_field_hash_instruction(avm_instrs, function, destinations, inputs),
+        _ => handle_getter_instruction(avm_instrs, function, destinations, inputs),
+    }
+}
+
+fn handle_2_field_hash_instruction(
+    avm_instrs: &mut Vec<AvmInstruction>,
+    function: &String,
+    destinations: &[ValueOrArray],
+    inputs: &[ValueOrArray],
+) {
+    // handle field returns differently
+    let hash_offset_maybe = inputs[0];
+    println!("hash_offset_maybe: {:?}", hash_offset_maybe);
+    let (hash_offset, hash_size) = match hash_offset_maybe {
+        ValueOrArray::HeapArray(HeapArray { pointer, size }) => (pointer.0, size),
+        _ => panic!("Keccak | Sha256 address inputs destination should be a single value"),
+    };
+
+    assert!(destinations.len() == 1);
+    let dest_offset_maybe = destinations[0];
+    let dest_offset = match dest_offset_maybe {
+        ValueOrArray::HeapArray(HeapArray { pointer, size }) => {
+            assert!(size == 2);
+            pointer.0
+        }
+        _ => panic!("Keccak | Poseidon address destination should be a single value"),
+    };
+
+    let opcode = match function.as_str() {
+        "keccak256" => AvmOpcode::KECCAK,
+        "sha256" => AvmOpcode::SHA256,
+        _ => panic!(
+            "Transpiler doesn't know how to process ForeignCall function {:?}",
+            function
+        ),
+    };
+
+    avm_instrs.push(AvmInstruction {
+        opcode,
+        indirect: Some(3), // 11 - addressing mode, indirect for input and output
+        operands: vec![
+            AvmOperand::U32 {
+                value: dest_offset as u32,
+            },
+            AvmOperand::U32 {
+                value: hash_offset as u32,
+            },
+            AvmOperand::U32 {
+                value: hash_size as u32,
+            },
+        ],
+        ..Default::default()
+    });
+}
+
+fn handle_field_hash_instruction(
+    avm_instrs: &mut Vec<AvmInstruction>,
+    function: &String,
+    destinations: &[ValueOrArray],
+    inputs: &[ValueOrArray],
+) {
+    // handle field returns differently
```

Comment:
Same as other comment:
"Should this be name `emit_*` like fcarreiro's functions here and made to return an instruction?

Also this is not trivial and could use a meaningful `///` comment prior to the fn regarding what this function is doing."